File: platform-nest/src/events/hlc.ts
// Hybrid Logical Clock (sync-engine revision §2, D3 fix #2 and #4). THE only clock stamped on
// every outbox_events row for cross-site ordering. Mirrors sync-engine-go/internal/hlc so both
// sides interoperate byte-for-byte.
//
// Canonical text format: ZERO-PADDED "%013d.%010d" (wallMs.counter). Padding makes plain text
// ordering equal logical ordering, so SQL `hlc > cursor` and `MAX(hlc)` are correct as text
// comparisons — see migrations/0012_outbox_hlc.sql.
//
// D3 #4 (failover monotonicity): seedFromPersisted() lifts the clock to at least the last HLC
// persisted for this origin_site, so a process restart (or a promoted standby with a lagging
// wall clock) can never mint an HLC that regresses behind what is already committed.

import type { Pool, PoolClient } from "pg";
import { config } from "../config";

const WALL_WIDTH = 13;
const CTR_WIDTH = 10;

export function formatHlc(wallMs: number, counter: number): string {
  return `${String(wallMs).padStart(WALL_WIDTH, "0")}.${String(counter).padStart(CTR_WIDTH, "0")}`;
}

export function parseHlc(s: string): { wallMs: number; counter: number } {
  const dot = s.indexOf(".");
  if (dot < 0) throw new Error(`invalid HLC string: ${s}`);
  return { wallMs: Number(s.slice(0, dot)), counter: Number(s.slice(dot + 1)) };
}

export class HlcClock {
  private lastMs = 0;
  private counter = 0;

  constructor(private readonly now: () => number = () => Date.now()) {}

  /** Monotonic next HLC. Same-wall-ms calls bump the counter; a wall advance resets it. */
  next(): string {
    const wall = this.now();
    if (wall > this.lastMs) {
      this.lastMs = wall;
      this.counter = 0;
    } else {
      this.counter++;
    }
    return formatHlc(this.lastMs, this.counter);
  }

  /** D3 #4 startup guard: never issue an HLC <= lastKnown. */
  seedFromPersisted(lastKnown: string | null | undefined): void {
    if (!lastKnown) return;
    const { wallMs, counter } = parseHlc(lastKnown);
    if (wallMs > this.lastMs || (wallMs === this.lastMs && counter > this.counter)) {
      this.lastMs = wallMs;
      this.counter = counter;
    }
  }
}

// Process-wide singleton (Node is single-threaded per process; no mutex needed for the
// synchronous next()). emitEvent() stamps every row from this instance.
let singleton: HlcClock | null = null;
export function getClock(): HlcClock {
  if (!singleton) singleton = new HlcClock();
  return singleton;
}

/**
 * Seed the singleton from the DB at startup (call from main bootstrap). Reads the greatest HLC
 * this origin_site has ever written, parsing numerically so it is robust regardless of format.
 */
export async function seedClockFromDb(db: Pool | PoolClient): Promise<void> {
  const { rows } = await db.query<{ hlc: string | null }>(
    `SELECT hlc FROM outbox_events
     WHERE origin_site = $1 AND hlc IS NOT NULL
     ORDER BY (split_part(hlc, '.', 1))::bigint DESC, (split_part(hlc, '.', 2))::bigint DESC
     LIMIT 1`,
    [config.originSite],
  );
  getClock().seedFromPersisted(rows[0]?.hlc ?? null);
}

File: sync-engine-go/internal/hlc/hlc.go
// Hybrid Logical Clock — THE only clock for cross-site ordering (sync-engine-revision §2,
// D3 fix #2: never updated_at). Byte-for-byte interoperable with platform-nest/src/events/hlc.ts.
//
// Canonical text format is ZERO-PADDED "%013d.%010d" (wallMs.counter). Padding makes plain text
// ordering equal logical ordering, so SQL `hlc > cursor` and `MAX(hlc)` are correct as text
// comparisons (unpadded "%d.%d" sorts "1000..." before "999..."). Compare() still uses the
// numeric fields; only the string form is padded.
//
// SeedFromPersisted implements D3 fix #4: on startup/promotion, lift the clock to at least the
// last HLC persisted for this origin_site so a promoted standby with a lagging wall clock can
// never mint an HLC that regresses behind what is already committed.
package hlc

import (
	"fmt"
	"strconv"
	"strings"
	"sync"
)

const (
	wallWidth = 13
	ctrWidth  = 10
)

type HLC struct {
	WallMs  int64
	Counter int32
}

func (h HLC) Compare(other HLC) int {
	if h.WallMs != other.WallMs {
		if h.WallMs < other.WallMs {
			return -1
		}
		return 1
	}
	if h.Counter != other.Counter {
		if h.Counter < other.Counter {
			return -1
		}
		return 1
	}
	return 0
}

func (h HLC) String() string {
	return fmt.Sprintf("%0*d.%0*d", wallWidth, h.WallMs, ctrWidth, h.Counter)
}

func Parse(s string) (HLC, error) {
	parts := strings.SplitN(s, ".", 2)
	if len(parts) != 2 {
		return HLC{}, fmt.Errorf("invalid HLC string: %q", s)
	}
	wall, err := strconv.ParseInt(parts[0], 10, 64) // ParseInt tolerates leading zeros
	if err != nil {
		return HLC{}, err
	}
	counter, err := strconv.ParseInt(parts[1], 10, 32)
	if err != nil {
		return HLC{}, err
	}
	return HLC{WallMs: wall, Counter: int32(counter)}, nil
}

type Clock struct {
	mu      sync.Mutex
	now     func() int64
	lastMs  int64
	counter int32
}

func NewClock(now func() int64) *Clock {
	return &Clock{now: now}
}

func (c *Clock) Next() HLC {
	c.mu.Lock()
	defer c.mu.Unlock()
	wall := c.now()
	if wall > c.lastMs {
		c.lastMs = wall
		c.counter = 0
	} else {
		c.counter++
	}
	return HLC{WallMs: c.lastMs, Counter: c.counter}
}

// SeedFromPersisted enforces the startup guard: the clock never issues an HLC <= lastKnown.
func (c *Clock) SeedFromPersisted(lastKnown HLC) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if lastKnown.WallMs > c.lastMs || (lastKnown.WallMs == c.lastMs && lastKnown.Counter > c.counter) {
		c.lastMs = lastKnown.WallMs
		c.counter = lastKnown.Counter
	}
}
